manager/orchestrator/replicatedjob/reconciler.go

```diff
@@ -0,0 +1,199 @@
+package replicatedjob
+
+import (
+	"fmt"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/orchestrator"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+// reconciler is the type that holds the reconciliation logic for the
+// orchestrator. It exists so that the logic of actually reconciling and
+// writing to the store is separated from the orchestrator, to make the event
+// handling logic in the orchestrator easier to test.
+type reconciler interface {
+	ReconcileService(id string) error
+}
+
+type reconcilerObj struct {
+	// we need the store, of course, to do updates
+	store *store.MemoryStore
+
+	// a copy of the cluster is needed, because we need it when creating tasks
+	// to set the default log driver
+	cluster *api.Cluster
+}
+
+// newReconciler creates a new reconciler object
+func newReconciler(store *store.MemoryStore, cluster *api.Cluster) reconciler {
+	return &reconcilerObj{
+		store:   store,
+		cluster: cluster,
+	}
+}
+
+// ReconcileService reconciles the replicated job service with the given ID by
+// checking to see if new replicas should be created. reconcileService returns
+// an error if there is some case prevent it from correctly reconciling the
+// service.
+func (r *reconcilerObj) ReconcileService(id string) error {
+	var (
+		service *api.Service
+		tasks   []*api.Task
+		viewErr error
+	)
+	// first, get the service and all of its tasks
+	r.store.View(func(tx store.ReadTx) {
+		service = store.GetService(tx, id)
+
+		tasks, viewErr = store.FindTasks(tx, store.ByServiceID(id))
+	})
+
+	if service == nil {
+		return nil
+	}
+
+	// errors during view should only happen in a few rather catastrophic
+	// cases, but here it's not unreasonable to just return an error anyway.
```

Comment:
Should arguably happen before the `if service == nil` check, actually?

Reply:
Yes, I think you're right.